### ISSUE TITLE
[rewrite] Change proposal: vnode.attrs => vnode.input

### ISSUE THREAD

**Issue by barneycarroll** [lhorie/mithril.js#1405]
When explaining & debugging component logic I always refer to the values supplied to the component in the second argument of `m()` as 'input'. I'd like to bikeshed the idea of formally renaming `attrs` to `input`.

# Why attrs?

'Attributes' is intuitive if we consider components to be analogous to custom elements: for people unfamiliar or dubious about the idea of declarative components sitting alongside vanilla DOM element declarations — especially in MSX — then the author's experience of working with HTML will make hyperscript very easy to grok.

# But it's kinda complicated

One of the more recondite difficulties of working with DOM scripting is the subtle distinction between attributes and properties. Like other virtual DOM libs, Mithril glosses over the internal logic determining whether any given property supplied in the second argument of an element hyperscript declaration is to be treated as a property or an attribute. Libraries that require the author to be explicit about the distinction are for the most part obnoxious, especially since coming from a background of static HTML, attributes and properties are indistinguishable. In this sense 'attrs' for attributes is a better choice than eg React's 'props' for properties.

# Thinking about it the wrong way

The fact that a hash of properties supplied as the second argument to `m()` looks like (and usually results in) attribute assignment is obvious to the eye and doesn't need reinforcing with a naming convention. The irony is that in the cases where the vnode is an element and `vnode.attrs` do end up representing DOM attributes, this is handled internally at the patching level and the author will likely never have cause to reference the `attrs` property — it's an implementation detail that doesn't require author intervention.

# The case for semantic distinction 

There's a holistic API design argument to be made that the automatic assignment of attributes to a DOM node via the second argument is one of many outcomes. To be explicit, the properties of the object can end up being any of the following:
* A DOM attribute assignment
* A DOM property assignment
* A lifecycle method
* Arbitrary values to be used as the author sees fit within lifecycle hooks and / or a component view 

So on a purely pedantic semantic level, it would be good to distinguish this part of the vnode API from the most common outcome.

# Missed opportunities in cognitive implications

By making an implicit equivalence of hyperscript input signature, the vnode object interface, and the declarative HTML attribute declaration convention, authors miss out on the potential of the 3 other use cases for hyperscript input listed above. People tend not to have a problem thinking of lifecycle hooks and DOM properties as 'special kinds of attributes' — that's a fairly harmless semantic error. But in the case of the last use case — authors writing components with their own esoteric input signatures and internal logic — the `attrs` conflation becomes harmful.

# How `input` would help

Given that DOM attribute / property assignment is totally intuitive and lifecycle method input is a documented (and arguably recondite) feature, the potential for powerful component design via author-designed input and internal logic is the feature that really unlocks the power of Mithril, and this should be emphasised.

I propose that the implicit emphasis on declarative direct assignment acts to the detriment of the powerful functional programming features Mithril has always excelled at revealing to its adopters. Preact and Inferno have stimulated people by emphasising functional attitudes to components, where components are functions which react (haha) to input over the course of their existence. Consider that lifecycle methods such as before / on update are largely meaningless without being able to react to new input and take according action. The risk is that people see 'attribute' definition as a largely static thing for the purposes of one time 'configuration' of their components on instantiation. This ellides the immense potential of dynamic input that changes over time, which is what makes dynamic virtual DOM so powerful as an idiom.

If we use `vnode.input`, I believe we can reinforce the idea of components as functions with reactive dynamic capability. This would, I believe, reduce the tendency for authors to get tied up in object-oriented messes where they believe components need to define all sorts of esoteric internal complexity in order to achieve dynamic functionality and communication patterns, as a result of believing attributes are a static initialisation feature. Essentially, I believe using the term `input` will make life easier for people trying to develop dynamic components and emphasises the value of lifecycle hooks and significant logic in views.

# Other small wins 

Attributes are harmful in the sense that people believing that a components input is essentially analogous to a custom element's attributes is limited to a hash of key value pairs. In many ways this relegates Mithril components to the perceived capability of W3C custom elements or Web components — whereas engaged Mithril users benefit from far more power in custom component input. Consider the ability to pass in a solitary stream or function as the input — or a deep object with custom types rather than a simple one dimensional hash. By broadening the nominal implications of `input`, users are empowered to devise their own arbitrary architectures and APIs while using the vnode core API to its fullest.

# TL;DR 

`attrs` overemphasises a fallacious cognitive dissonance that may be comforting to unambitious adopters who may be glad to see a familiar term, but `input` presents a more accurate distinction in terminology which encourages ambition and suggests potential in a way that aligns with Mithril's most rewarding virtues.

**Comment by isiahmeadows** [comment]
I appreciate the point of view, but I feel `input` is not much of an
improvement.

A program accepts input and emits output. A process accepts input and emits
output. It performs actions, but is not an action itself. It can figure out
things around them, using components and data to remember things
temporarily, but they know nothing beyond what they're taught to observe.

A function accepts parameters/arguments and returns zero or more values. It
is an action, and may return the result of its action. It also knows
nothing, because actions have no memory other than the last thing they've
done. They can observe what you tell them to, but you also have to tell
them to store it somewhere.

A component has properties and children. An interface has properties. An
element has attributes and children. A record has keys/attributes. These
cannot perform actions, nor are they actions. They're stateful nouns, with
no ability to act or even know (the GoF only pretends they can). You can
observe qualities of them (like reading a property), but observation is an
action itself. They have no capability to act because they are in a sense,
imaginary.

Let's not conflate these three. The God object is the perfect example of
why you shouldn't:

1. It tries to both remember all state and have ways of manipulating its
state, confusing programs with components.
2. It tries to provide ways to manipulate itself while observing
everything, confusing programs with actions.
3. It has both a memory and ways to manipulate it indirectly, confusing
components with actions.

Why this long, big, writeup? For this one reason: if we're going to change
something for philosophical reasons, we should actually understand why a)
the current state is why it is, and b) have a sound philosophical basis for
the change as well.

As for the change, I like the idea, but `input` is arguably worse than the
status quo. Also, don't forget that @lhorie wants to get a release out the
door.

**Comment by lhorie** [comment]
-1 from me

The rewrite has been in development for a long time, and at this stage, API changes are going to hold back the release.
I feel that the cost/benefit ratio for this change is not favorable towards implementing it (or to put it bluntly, it doesn't really accomplish enough to justify forcing people to rename .attrs in their code.

On the topic itself: I don't feel `input` is a good name. That term typically refers to `<input>` tags. Also, while it makes a little bit of sense to think in terms of input and output when you're talking in the context of running procedural code in a function, the issue here is that `vnode.attrs` is not the whole input, `vnode` is. (Or, to be more pedantic, the arguments of the function are the input). 

Vnode is a data structure, so its fields should be descriptive of what they represent in the context of what the data structure represents. React uses `props` because that's what they represent: `class` is not a valid property of Element; `className` is. Mithril uses `attrs` because imho that's what makes more sense. Nobody wants to memorize case sensitivity exceptions in field names if they can help it. Can Mithril `attrs` technically be properties, lifecycle hooks and custom arguments? Yes, but when you think of a vnode, you typically understand what it is by mapping it to an HTML element, and when you think of HTML elements, you think of attributes, not properties or arguments or hooks or inputs.
